components/properties/src/provider.rs

```diff
@@ -311,11 +317,140 @@ impl<'data, T: TrieValue> PropertyCodePointMapV1<'data, T> {
+        }
+    }
+}
+
+#[cfg(feature = "serde")]
+impl<'de> serde::Deserialize<'de> for Box<NormalizedPropertyNameStr> {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        use alloc::borrow::Cow;
+        let s; // lifetime lengthening
+        let b;
+        // Can be improved with https://github.com/unicode-org/icu4x/issues/2310
+        // the allocations here are fine, in normal ICU4X code they'll only get hit
+        // during human-readable deserialization
+        let bytes = if deserializer.is_human_readable() {
+            s = <Cow<str>>::deserialize(deserializer)?;
+            s.as_bytes()
+        } else {
+            b = <Cow<[u8]>>::deserialize(deserializer)?;
+            &b
+        };
+        Ok(NormalizedPropertyNameStr::from_bytes(bytes))
+    }
+}
+
+impl<'a> ZeroMapKV<'a> for NormalizedPropertyNameStr {
+    type Container = VarZeroVec<'a, NormalizedPropertyNameStr>;
+    type Slice = VarZeroSlice<NormalizedPropertyNameStr>;
+    type GetType = NormalizedPropertyNameStr;
+    type OwnedType = Box<NormalizedPropertyNameStr>;
+}
+
+impl PartialOrd for NormalizedPropertyNameStr {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+/// Normalize a character based on the "loose matching" described in PropertyValueAliases.txt,
+/// returning None for skippable characters
+///
+/// ICU has [code for this][1] (and [during property lookup][2]) which we emulate.
+/// In particular, ICU only does normalization within ASCII, which makes sense since character names
+/// seem to be only ASCII.
+///
+/// [1]: https://github.com/unicode-org/icu/blob/288c4c7555915ce7b1fb675d94ddd495058fc039/icu4c/source/common/propname.cpp#L35
+/// [2]: https://github.com/unicode-org/icu/blob/288c4c7555915ce7b1fb675d94ddd495058fc039/icu4c/source/common/propname.cpp#L226-L230
+fn normalize_char(ch: u8) -> Option<u8> {
+    match ch {
+        b'a'..=b'z' => Some(ch),
+        b'A'..=b'Z' => Some(ch | 0x20), // Uppercase by setting last bit
+        // underscores, hyphens, and all ASCII whitespace
+        b'_' | b'-' | b' ' | 0x09..=0x0d => None,
+        _ => Some(ch),
+    }
+}
+
+impl Ord for NormalizedPropertyNameStr {
+    fn cmp(&self, other: &Self) -> Ordering {
+        let self_iter = self.0.iter().copied().filter_map(normalize_char);
+        let other_iter = other.0.iter().copied().filter_map(normalize_char);
+        self_iter.cmp(other_iter)
+    }
+}
+
+impl fmt::Debug for NormalizedPropertyNameStr {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        if let Ok(s) = str::from_utf8(&self.0) {
+            f.write_str(s)
+        } else {
+            f.write_str("(invalid utf8)")
+        }
+    }
+}
+
+impl NormalizedPropertyNameStr {
+    #[cfg(feature = "serde")]
+    /// Get a Box<NormalizedPropertyName> from a byte slice
+    pub fn from_bytes(b: &[u8]) -> Box<Self> {
```

Comment:
Nit: the function named `from_bytes` should return `&Self` not `Box<Self>`. Suggestion: use `from_boxed_bytes` and don't allocate memory

Reply:
that needs a transmute

Reply:
Also we need this in datagen in a context where we don't have something boxed